src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7367,10 +7367,43 @@ bool canBeNullableType()
                             // More expensive tests are in `EatNullableQualifierIfApplicable`
                             if (type.Kind == SyntaxKind.NullableType || type.Kind == SyntaxKind.PointerType)
                                 return false;
+
                             if (this.PeekToken(1).Kind == SyntaxKind.OpenBracketToken)
+                            {
+                                // T?[
+                                //
+                                // This could be a array of nullable types (e.g. `is T?[]` or `is T?[,]`) or it's a
+                                // conditional with a collection expression or lambda (e.g. `is T ? [...] :` or `is T ? [Attr]() => ...`)
+                                //
+                                // Note: `is T?[]` could be the start of either.  So we have to look to see if we have a
+                                // `:` to know which case we're in.
+                                if (mode == ParseTypeMode.AfterIs)
+                                {
+                                    switch (this.PeekToken(2).Kind)
+                                    {
+                                        // `is T?[,]`.  Definitely an array of nullable type.
+                                        case SyntaxKind.CommaToken:
+                                            return true;
+
+                                        // `is T?[]`.  Could be an array of a nullable type, or a conditional.  Have to
+                                        // see if it is followed by `:` to find out.  If there is a colon, it's a
+                                        // conditional.
+                                        case SyntaxKind.CloseBracketToken:
+                                            return this.PeekToken(3).Kind != SyntaxKind.ColonToken;
```

Comment:
If collection-exprs become usable as a primary-expression, we'll have to revisit this heuristic, right?
e.g. for `x is T ? [].Method() : ...`

Also, are there any potentially valid expression forms today, which don't have the collection immediately followed by `:`, like `x is T ? [] == GetCollection() : ...`?

Reply:
> If collection-exprs become usable as a primary-expression, we'll have to revisit this heuristic, right?

Yes.  We'd need to revisit that based on what can legally follow.   Fun times.

> Also, are there any potentially valid expression forms today, which don't have the collection immediately followed by :, like x is T ? [] == GetCollection() : ...?

THis is definitely an interesting case.  I'm thinking we may actually have to speculatively parse here to see what is happening.